[WTF] Provide a type hint to debuggers for opaque bitfield values where supported

weinig opened the pull request:
#### 99b626befab865c4f50048ef2c567c2a10d5d053
<pre>
[WTF] Provide a type hint to debuggers for opaque bitfield values where supported
<a href="https://bugs.webkit.org/show_bug.cgi?id=295294">https://bugs.webkit.org/show_bug.cgi?id=295294</a>

Reviewed by Darin Adler.

Newer versions of clang support an attribute, `preferred_type`, which provides
a hint to debuggers about what type is stored in opaqued bitfields:

    <a href="https://clang.llvm.org/docs/AttributeReference.html#preferred-type.">https://clang.llvm.org/docs/AttributeReference.html#preferred-type.</a>

This adds a new macro to wtf/Compiler.h called `PREFERRED_TYPE(...)` that
wraps the attribute when supported.

Some initial adoption of it was done in &apos;WebCore/rendering/style&apos; to
illustrate how it can be used.

* Source/WTF/wtf/Compiler.h:
    - Adds new PREFERRED_TYPE macro.
    - Cleans up Compiler.h a bit, fixing typos, reducing some duplication and
      making COMPILER_HAS_CPP_ATTRIBUTE a top level macro.

* Source/WebCore/rendering/style/StyleBoxData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleMultiColData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
    - Adopted PREFERRED_TYPE, replacing comments annotations.

Canonical link: <a href="https://commits.webkit.org/297027@main">https://commits.webkit.org/297027@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea23298394ebaf5d1152ea842c1cd3ac504e95c5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110166 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29825 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20257 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116188 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60414 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112129 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30503 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38412 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83767 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113114 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99214 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64212 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23705 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17397 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59984 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/102657 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93717 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17407 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118979 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/108720 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37206 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27586 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92744 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37578 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95483 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92567 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23611 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37553 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15290 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33095 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37100 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42571 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132995 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36762 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35963 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40102 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38471 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -43,6 +43,15 @@
 #define COMPILER_HAS_ATTRIBUTE(x) 0
 #endif
 
+/* COMPILER_HAS_ATTRIBUTE() - whether the compiler supports a particular c++ attribute. */
```
Given the others in this file, I sort of would've expected `CLANG` to be in the name of this macro. Or maybe the right direction is to remove it from the other similar macros some day.

weinig replied:
Its supported by GCC as well, so I think the direction should be remove clang from the ones that don't need it.

`__has_builtin` is also supported by GCC (https://gcc.gnu.org/onlinedocs/cpp/_005f_005fhas_005fbuiltin.html) so it should definitely be updated.

darinadler commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -183,7 +192,7 @@
 /* ALWAYS_INLINE */
 
 /* In GCC functions marked with no_sanitize_address cannot call functions that are marked with always_inline and not marked with no_sanitize_address.
- * Therefore we need to give up on the enforcement of ALWAYS_INLINE when bulding with ASAN. https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67368 */
```
Did you grep for this typo through the whole source tree?

weinig replied:
heh, no, just saw it here and couldn't resist.

weinig replied:
But, doing so now, I see I got them all!

darinadler commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -268,16 +277,10 @@
-#if __SIZEOF_POINTER__ == 8
-#if !defined(MUST_TAIL_CALL) && defined(__cplusplus) && defined(__has_cpp_attribute)
-#if __has_cpp_attribute(clang::musttail) && !defined(__powerpc__) && !defined(_WIN32)
+#if !defined(MUST_TAIL_CALL) && COMPILER(CLANG) && __SIZEOF_POINTER__ == 8 && defined(__cplusplus) && COMPILER_HAS_CPP_ATTRIBUTE(clang::musttail) && !defined(__powerpc__) && !defined(_WIN32)
```
Not new, but seems like overkill to do both `COMPILER(CLANG)` and `COMPILER_HAS_CPP_ATTRIBUTE(clang::musttail)`.

darinadler commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -557,11 +567,11 @@
 #endif
 
 /* NOESCAPE */
-/* This attribute promises that a function argumemnt will only be used for the duration of the function,
```
Did you grep for this typo in the whole source tree?

weinig replied:
Just did, and this is the only one.

darinadler commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -614,18 +622,10 @@
 
 /* WTF_UNSAFE_BUFFER_USAGE */
 
-#ifndef __has_attribute
-#define __has_attribute(x) 0
-#endif
-
-#ifndef __has_cpp_attribute
-#define __has_cpp_attribute(x) 0
-#endif
```
Your new way is better.

darinadler commented on Source/WebCore/rendering/style/StyleBoxData.h:
```diff
@@ -91,11 +91,11 @@ class StyleBoxData : public RefCounted<StyleBoxData> {
-    unsigned m_boxSizing : 1; // BoxSizing
-    unsigned m_boxDecorationBreak : 1; // BoxDecorationBreak
-    unsigned m_verticalAlign : 4; // VerticalAlign
+    PREFERRED_TYPE(bool) unsigned m_hasAutoSpecifiedZIndex : 1;
+    PREFERRED_TYPE(bool) unsigned m_hasAutoUsedZIndex : 1;
+    PREFERRED_TYPE(BoxSizing) unsigned m_boxSizing : 1;
+    PREFERRED_TYPE(BoxDecorationBreak) unsigned m_boxDecorationBreak : 1;
+    PREFERRED_TYPE(VerticalAlign) unsigned m_verticalAlign : 4;
```
Almost wish the `unsigned` was part of the macro too, just for brevity. Maybe instead of a `PREFERRED_TYPE` macro there could be a `TYPED_BITFIELD` macro, but I guess that name over promises a bit.

weinig replied:
One day, c++ reflection will allow us to build the type casting getter/setters too, but not this day.

weinig committed:
[WTF] Provide a type hint to debuggers for opaque bitfield values where supported
https://bugs.webkit.org/show_bug.cgi?id=295294

Reviewed by Darin Adler.

Newer versions of clang support an attribute, `preferred_type`, which provides
a hint to debuggers about what type is stored in opaqued bitfields:

    https://clang.llvm.org/docs/AttributeReference.html#preferred-type.

This adds a new macro to wtf/Compiler.h called `PREFERRED_TYPE(...)` that
wraps the attribute when supported.

Some initial adoption of it was done in 'WebCore/rendering/style' to
illustrate how it can be used.

* Source/WTF/wtf/Compiler.h:
    - Adds new PREFERRED_TYPE macro.
    - Cleans up Compiler.h a bit, fixing typos, reducing some duplication and
      making COMPILER_HAS_CPP_ATTRIBUTE a top level macro.

* Source/WebCore/rendering/style/StyleBoxData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleMultiColData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleSelfAlignmentData.h:
    - Adopted PREFERRED_TYPE, replacing comments annotations.

Canonical link: https://commits.webkit.org/297027@main
